go-controller/pkg/clustermanager/status_manager/status_manager.go

```diff
@@ -66,50 +89,105 @@ func (m *typedStatusManager[T]) Stop() {
 	m.objController.Stop()
 }
 
-func (m *typedStatusManager[T]) needsUpdate(oldObj, newObj *T) bool {
-	if oldObj == nil || newObj == nil {
-		return true
-	}
-	return m.resource.statusChanged(oldObj, newObj)
-}
-
 func (m *typedStatusManager[T]) updateStatus(key string) error {
 	namespace, name, err := cache.SplitMetaNamespaceKey(key)
 	if err != nil {
-		klog.Errorf("Failed to split meta namespace cache key %s: %v", key, err)
+		klog.Errorf("StatusManager %s: failed to split meta namespace cache key %s for %s: %v", m.name, key, err)
 		return nil
 	}
 	obj, err := m.resource.get(namespace, name)
 	if err != nil {
-		return fmt.Errorf("expecting %T but received %T", *new(T), obj)
+		return fmt.Errorf("StatusManager %s: expecting %T but received %T", m.name, *new(T), obj)
 	}
 
-	applyOptions := &metav1.ApplyOptions{
-		Force:        true,
-		FieldManager: clusterManagerName,
+	return m.withZonesRLock(func(zones sets.Set[string]) error {
+		if zones.Has(zone_tracker.UnknownZone) || zones.Len() == 0 {
+			// zones are not in a consistent state, wait for more information to be populated
+			return nil
+		}
+
+		messages := m.resource.getMessages(obj)
+		if len(messages) == 0 {
+			return nil
+		}
+
+		// first, make sure no stale zones are present.
+		// every zone has exactly 1 status message
+		if len(messages) > zones.Len() {
+			for _, message := range messages {
+				if zoneID := types.GetZoneFromStatus(message); !zones.Has(zoneID) {
+					// stale zone, remove
+					// use zoneID as fieldManager to reset fields owned by that zone
+					applyAsZoneController := &metav1.ApplyOptions{
+						Force:        true,
+						FieldManager: zoneID,
+					}
+					klog.Infof("StatusManager %s: delete stale zone %s", m.name, zoneID)
+					err = m.resource.cleanupStatus(obj, applyAsZoneController)
+					if err != nil {
+						return err
+					}
+				}
+			}
+		}
+
+		// now calculate accumulated status.
+		// if not all zones reported status, clean it up, since the status is considered unknown until all zone report results.
```

Comment:
yeah we can set failed status without having all zone messages, I just didn't think it is important enough to reconcile for that case, but I can add it

Reply:
I also didnt want to make resourceManagers zone-aware, so I will need to pass an extra flag to only apply status if it is failure

Reply:
Yeah, since it is private to the package it is not a big deal. I guess it wouldn't be a big deal either passing the zones that the resourceManagers should expect a message from, so they filter out messages that should not be taken into account.